### PR TITLE
Add code coverage to grcov

### PR DESCRIPTION
**Pull request by calixteman**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -112,6 +115,14 @@ script:
     - $CXX --version
     - if [ -z "$TRAVIS_TAG" ]; then cargo build --verbose; fi
     - if [ -z "$TRAVIS_TAG" ]; then cargo test -- --nocapture; fi
+    - |
+      if [ -z "$COVERAGE" ]; then
+        cargo clean;
+        CXXFLAGS="--coverage" LDFLAGS="--coverage -lgcov" CARGO_INCREMENTAL=0 RUSTFLAGS="-Zprofile -Ccodegen-units=1" cargo +nightly test;
```
Comment (marco-c) [inline .travis.yml]: We can run the tests with these directly, so we don't need to run the tests twice

File: .travis.yml [hunk 2]
```diff
@@ -112,6 +115,14 @@ script:
...
+        zip -0 ccov.zip `find . \( -name "grcov*.gc*" -o -name "llvmgcov.gc*" \) -print`;
+        ./target/debug/grcov ccov.zip -s src/ -t lcov --llvm --branch --ignore-dir libstd/* --ignore-dir libcore/* --ignore-dir liballoc/* --ignore-dir libtest/* --ignore-dir "<*" --ignore-dir "path_rewriting.rs" > lcov.info;
```
Comment (marco-c) [inline .travis.yml]: Why do we need to ignore `path_rewriting.rs`?

File: .travis.yml [hunk 3]
```diff
@@ -95,14 +95,17 @@ matrix:
         - GCOV=gcov-7
         - CLANG_CXX=clang++-7
         - LLVM_CONFIG=llvm-config-7
+        - COVERAGE=1
```
Comment (marco-c) [inline .travis.yml]: Instead of defining COVERAGE, we can just upload coverage whenever we're using Rust nightly.

Comment (marco-c) [reply .travis.yml]: This way we also merge the coverage we get in the different compiler configurations

File: .travis.yml (hunk 2 above)
Comment (calixteman) [reply .travis.yml]: It's strange here, in fact one gcno contains the path path_rewriting.rs and src/path_rewriting.rs.
So the latter is correct but not the former.

File: .travis.yml [hunk 4]
```diff
@@ -103,15 +103,25 @@ matrix:
...
     - if [ "$TRAVIS_OS_NAME" = "osx" ]; then brew cask uninstall --force oclint; fi
     - if [ "$TRAVIS_OS_NAME" = "osx" ]; then brew install llvm@7 && brew link --overwrite --force llvm@7; fi
 script:
     - export CXX=$CLANG_CXX
     - $CXX --version
-    - if [ -z "$TRAVIS_TAG" ]; then cargo build --verbose; fi
-    - if [ -z "$TRAVIS_TAG" ]; then cargo test -- --nocapture; fi
+    - |
+      if [ "$TRAVIS_RUST_VERSION" = "nightly" ] && [ -z "$TRAVIS_TAG" ]; then
+        CXXFLAGS="--coverage" LDFLAGS="--coverage -lgcov" CARGO_INCREMENTAL=0 RUSTFLAGS="-Zprofile -Ccodegen-units=1" cargo +nightly build --verbose;
```
Comment (marco-c) [inline .travis.yml]: Nit: you could do something like:
```
- if [ "$TRAVIS_RUST_VERSION" = "nightly" ] && [ -z "$TRAVIS_TAG" ]; then export ... ; fi
- cargo build --release
- cargo test --nocapture
- if [ "$TRAVIS_RUST_VERSION" = "nightly" ] && [ -z "$TRAVIS_TAG" ]; then PARSE AND UPLOAD; fi

File: .travis.yml (hunk 2 above)
Comment (marco-c) [reply .travis.yml]: Weird, maybe it's worth filing another bug in the Rust repository about this.

**Pull request merged**